File: common/src/main/java/com/ssblur/scriptor/word/action/ExplosionAction.java
package com.ssblur.scriptor.word.action;

import com.ssblur.scriptor.api.word.Action;
import com.ssblur.scriptor.api.word.Descriptor;
import com.ssblur.scriptor.helpers.targetable.Targetable;
import com.ssblur.scriptor.word.descriptor.power.StrengthDescriptor;
import net.minecraft.server.level.ServerLevel;
import net.minecraft.world.level.Level;

public class ExplosionAction extends Action {
  @Override
  public void apply(Targetable caster, Targetable targetable, Descriptor[] descriptors) {
    if(targetable.getLevel().isClientSide) return;
    double strength = 2;
    for(var d: descriptors) {
      if(d instanceof StrengthDescriptor strengthDescriptor)
        strength += (float) strengthDescriptor.strengthModifier();
    }

    ServerLevel level = (ServerLevel) targetable.getLevel();
    var pos = targetable.getTargetPos();

    float power = (float) Math.pow(strength, 0.7);

    level.explode(null, pos.x, pos.y + .25, pos.z, power, Level.ExplosionInteraction.TNT);
  }

  @Override
  public Cost cost() { return new Cost(16, COSTTYPE.ADDITIVE); }

}

File: common/src/main/kotlin/com/ssblur/scriptor/word/action/ExplosionAction.kt
package com.ssblur.scriptor.word.action

import com.ssblur.scriptor.api.word.Action
import com.ssblur.scriptor.api.word.Descriptor
import com.ssblur.scriptor.helpers.targetable.EntityTargetable
import com.ssblur.scriptor.helpers.targetable.Targetable
import com.ssblur.scriptor.word.descriptor.power.StrengthDescriptor
import net.minecraft.server.level.ServerLevel
import net.minecraft.world.entity.Entity
import net.minecraft.world.level.Explosion
import net.minecraft.world.level.ExplosionDamageCalculator
import net.minecraft.world.level.Level

class ExplosionAction : Action() {
    internal class ExplosionActionDamageCalculator(val caster: Targetable) : ExplosionDamageCalculator() {
        override fun getEntityDamageAmount(explosion: Explosion, entity: Entity): Float {
            if (caster is EntityTargetable && caster.targetEntity === entity) return super.getEntityDamageAmount(
                explosion,
                entity
            ) * 0.15f
            return super.getEntityDamageAmount(explosion, entity) * 0.3f
        }
    }

    override fun apply(caster: Targetable, targetable: Targetable, descriptors: Array<Descriptor>) {
        if (targetable.level.isClientSide) return
        var strength = 2.0
        for (d in descriptors) {
            if (d is StrengthDescriptor) strength += d.strengthModifier()
        }

        val level = targetable.level as ServerLevel
        val pos = targetable.targetPos

        val power = Math.pow(strength, 0.7).toFloat()

        level.explode(
            null,
            null,
            ExplosionActionDamageCalculator(caster),
            pos.x,
            pos.y + .25,
            pos.z,
            power,
            false,
            Level.ExplosionInteraction.TNT
        )
    }

    override fun cost(): Cost {
        return Cost(16.0, COSTTYPE.ADDITIVE)
    }
}
